managed-pointer/cpp-20/D0567R1.md

```diff
@@ -5,222 +5,384 @@
+### 2.1. Must support different memory architectures
 
-Alternative names for the `put()` and `get()` interface were `acquire()` and `release()` as you were effectively acquiring and releasing the managed memory allocation and `send()` and `receive()` as you are effectively sending and receiving back the managed memory allocation.
+In non-heterogeneous or non-distributed systems, there is typically a single device: a host CPU with a single memory region or an area of addressable memory. In contrast, heterogeneous and distributed systems have multiple devices (including the host CPU) often with their own discrete memory regions. A device in this respect can be any architecture that exists within a system that is C\+\+ programmable; this can include CPUs, GPGPUs, APUs, FPGAs, DSPs, NUMA nodes, I/O devices and other forms of accelerators.
+
+These memory regions can be categorised into two groups. Firstly memory architectures which have a single unfied address space, that is a single area of addressable memory that is accessible across multiple devices, this can optionally be supprted in hardware by devices having the same physical memory or in software by having a cache coherancy layer accessing a shared virtual address space. Secondly memory architectures which have multiple address spaces, therefore having separate discrete areas of addresable memory, often in different physical memory.
+
+A unified interface needs to be capable of supporting both of these kinds of memory architectures in such a way that any requirements for cache coherancy is abstracted by the interface and associated coherency model.
+
+### 2.2. Must support different levels of synchronisation
+
+Different memory architectures, both unified and discrete; may support granularity at different levels. This can either be at the course grained level; where synchronisation between memory regions is performed at the execution function level, or this can be at the fine grained level; where synchronisation between memory regions is performed at the instruction level, allowing access to memory from different devices concurrently, using atomics to specify read and write order.
+
+A unified interface needs to be capable of supporting both these levels of synchronisation.
 
-## Proposed Additions
+### 2.3. Must support different data flow models
 
-### Requirements on Execution Context
+There are also many different data flow models that can be used to describe data and data dependenies in relation to functions of execution. Though it would be impossible to cover all possible models, it's important to make a unified inteface as flexible and utilisable as possible, therfore facilitating a wide range of data flow models.
```

Comment:
I have expanded on this section to describe what is meant by data-flow model and the kinds of data-flow models that the proposal should support.